druid-shell/src/platform/web/window.rs

```diff
@@ -0,0 +1,633 @@
+// Copyright 2020 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Web window creation and management.
+
+use instant::Instant;
+use std::any::Any;
+use std::cell::{Cell, RefCell};
+use std::ffi::OsString;
+use std::rc::{Rc, Weak};
+use std::sync::{Arc, Mutex};
+
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+
+use crate::kurbo::{Point, Size, Vec2};
+
+use crate::piet::RenderContext;
+
+use super::error::Error;
+use super::menu::Menu;
+use crate::common_util::IdleCallback;
+use crate::dialog::{FileDialogOptions, FileDialogType, FileInfo};
+
+use crate::keyboard;
+use crate::keycodes::KeyCode;
+use crate::mouse::{Cursor, MouseButton, MouseEvent};
+use crate::window::{IdleToken, Text, TimerToken, WinHandler};
+use crate::KeyModifiers;
+
+use super::util::init_log;
+use log::{error, warn};
```

Comment:
I'm discouraging use of these macros directly; I think the code is clearer when they are qualified through the module, i.e. `log::error!("Oh no!")` instead of `error!("Oh no")`; the latter looks like it might panic or something.